narwhals/_compliant/group_by.py

```diff
@@ -67,14 +70,43 @@ def compliant(self) -> CompliantFrameT_co:
     def __init__(
         self,
         compliant_frame: CompliantFrameT_co,
-        keys: Sequence[str],
+        keys: Sequence[CompliantExprT_contra],
         /,
         *,
         drop_null_keys: bool,
     ) -> None: ...
 
     def agg(self, *exprs: CompliantExprT_contra) -> CompliantFrameT_co: ...
 
+    @staticmethod
+    def _magic_parsing(
+        compliant_frame: CompliantLazyFrameAny, keys: Sequence[CompliantExprT_contra]
+    ) -> tuple[CompliantLazyFrameAny, list[str], list[str]]:
+        plx = compliant_frame.__narwhals_namespace__()
+        suffix_token = "_" * max(len(str(c)) for c in compliant_frame.columns)
+        original_names = compliant_frame._evaluate_aliases(*[k.name.keep() for k in keys])
+        output_names = compliant_frame._evaluate_aliases(*keys)
+
+        safe_keys = [
+            key.name.suffix(suffix_token)
+            if key._metadata is not None
+            and key._metadata.expansion_kind is ExpansionKind.MULTI_NAMED
+            else key.alias(f"{new_name}{suffix_token}")
```

Comment:
If it's multinamed, then it cannot be duplicate, therefore as temporary name we can just use the original one with a suffix. Otherwise, we alias with a suffix